contrib/opentelemetry/temporal_header_carrier.go

```diff
@@ -0,0 +1,43 @@
+package opentelemetry
+
+import (
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/sdk/converter"
+)
+
+// TemporalHeaderCarrier adapts temporal header to satisfy the Carrier interface.
+type TemporalHeaderCarrier map[string]*commonpb.Payload
```

Comment:
Is this PR just to create this type alias and associate methods?

Reply:
apologies I was just about to create an issue for this, yes the PR is to add a Opentelemetry "carrier" implementation, it would be used like this:

```
import  (
    "go.opentelemetry.io/otel/propagation"
    "go.temporal.io/sdk/interceptor"
)

carrier := temporal_propagator.TemporalHeaderCarrier(interceptor.WorkflowHeader(ctx)

// In order to copy otel TextMap data from context into Temporal Headers
propagation.TextMapPropagator.Inject(ctx, carrier)

// In order to copy the otel TextMap data from Temporal Headers into context
propagation.TextMapPropagator.Extract(ctx, carrier)
```

Reply:
Carrier definition https://opentelemetry.io/docs/specs/otel/context/api-propagators/#carrier

Reply:
It can be added to the SDK but I was unable to find the place where the Trace ID gets copied to the headers, the reason it would be beneficial for the temporal interceptor to use this is that it makes the interceptors compatible with other opentelemetry instrumentations

i.e. if you have temporal interceptors installed, and from within a workflow you make an Otel instrumented grpc request, the trace ID will get propagated all the way through

Reply:
would you be able to point me to the interceptor code that deals with copying trace ID into the header? I can try to update it and get the tests working

Reply:
My suggestion was if you think others would find this useful, to open source it as its own library.

Reply:
the thing is, I am suggesting for the temporal interceptor to use this carrier, so that the temporal interceptors are compatible with other opentelemetry library instrumentations. Would I be better to raise an issue for this?

It would need to be used here (and some refactoring of the parent function) https://github.com/temporalio/sdk-go/blob/master/contrib/opentelemetry/tracing_interceptor.go#L146. Looks like @jvonfricken and @cretz worked on the original

Reply:
`TextMapPropagator` is already an option you can specify, What additional customization point do you need?

Reply:
so this request is about the carrier (not the propagator), the suggestion I have is to replace `[textMapCarrier](https://github.com/temporalio/sdk-go/blob/master/contrib/opentelemetry/tracing_interceptor.go#L146)` as the carrier here is actually the temporal header. We would like to be able to use a common function with the temporal.io otel interceptor, so that we can be guaranteed that any custom interceptors we write will use the same header keys as the temporal.io otel interceptor. 

It also allows for extensions if more data types are required to be sent in the future, they can easily be added to the temporal headers whilst aligning with Open telemetry patterns.

Reply:
fwiw I raised an issue here, in case it explains it better https://github.com/temporalio/sdk-go/issues/1535

Reply:
>the suggestion I have is to replace textMapCarrier [link](https://github.com/temporalio/sdk-go/blob/master/contrib/opentelemetry/tracing_interceptor.go#L146) as the carrier here is actually the temporal header

The carrier is just a payload in the header, not the whole header.

>so that we can be guaranteed that any custom interceptors we write will use the same header keys as the temporal.io otel interceptor.

I am not sure how replacing the carrier accomplishes your goal. The `HeaderKey` and `SpanContextKey` are options in `TracerOptions`

Reply:
For example if the temporal starter is not owned by us, we cannot specify the Header key that the interceptor uses. If the workers then have our own custom interceptor installed, we cannot read the trace data from that Header key as it is not exposed.

Reply:
> For example if the temporal starter is not owned by us

How can you ensure that it is using OTel interceptor at all? If you can ensure the starter will set the client options to use a certain interceptor, surely you can ensure the starter can set certain interceptor options?

Reply:
Yes whilst it is not guaranteed that the starter is using the otel interceptor, it does maximises compatibility if all interceptors are able to use the same carrier. If another upstream team (or company!) installs otel they are likely to use your interceptor and not likely to set their own HeaderKey as they dont know what the downstream will use. 

This is about maximising compatibility, if you wanted to expose the HeaderKey that you use that could also be a solution.

Reply:
> if you wanted to expose the HeaderKey that you use that could also be a solution.

We intentionally don't expose this because it is so easy to set and we don't want people to build a reliance upon it. This is why we make the option available.

Reply:
I understand what you are saying, unfortunately it doesnt help in our case where we do not control the starter service.